CreditManagement/models.py

```diff
@@ -1,196 +1,123 @@
-                'source_association': ValidationError('Either a user or association should be given as a source', code='invalid'),
-            })
-        if self.target() is None:
-            raise ValidationError({
-                'target_user': ValidationError('Either a user or association should be given as a target', code='invalid'),
-                'target_association': ValidationError('Either a user or association should be given as a target.', code='invalid'),
-            })
-        super(Transaction, self).clean()
+        """
+        Transaction business rules.
+        """
 
-    def __str__(self):
-        return "€"+str(self.amount) + " " + str(self.source()) + " to " + str(self.target())
+        # Balance bottom limit
+        if self.source_user:
+            balance = self.source_user.balance
+            new_balance = balance - self.amount
+            if new_balance < settings.MINIMUM_BALANCE:
+                raise ValidationError(_("Balance becomes too low"))
 
+        # Associations cannot transfer money between each other
+        if self.source_association and self.target_association:
+            raise ValidationError(_("Associations cannot transfer money between each other"))
 
-class AssociationCredit(models.Model):
-    association = models.ForeignKey(Association, on_delete=models.SET_NULL, null=True)
-    start_date = models.DateField(auto_now_add=True)
-    end_date = models.DateField(blank=True, null=True)
-    credit = models.DecimalField(verbose_name="Credit balance", decimal_places=2, max_digits=6, default=0)
-    isPayed = models.BooleanField(default=False)
-    isLocked = models.BooleanField(default=False)
+    def __str__(self):
+        return "{} | {} | {} → {} | {}".format(self.moment, self.amount, self.source(), self.target(), self.notes)
 
-    def save(self, *args, **kwargs):
-        """
-        Overwrite the save function to lock changes after closure
-        :param args: not used
-        :param kwargs:
-        :return: None
-        """
-        if self.end_date is not None:
-            if self.isLocked is False:
-                self.isLocked = True
-                super(AssociationCredit, self).save(*args, **kwargs)
-                AssociationCredit(association=self.association).save()
-                return
-
-        if self.isLocked:
-            super(AssociationCredit, self).save(update_fields=['isPayed'], **kwargs)
-            return
-        super(AssociationCredit, self).save(*args, **kwargs)
 
-    def __str__(self):
-        if self.end_date is not None:
-            return self.association.name + " [" + self.start_date.strftime('%x') + " - " + self.end_date.strftime('%x') + "]"
-        else:
-            return self.association.name + " [" + self.start_date.strftime('%x') + " - now ]"
+# Todo: this is not yet in use
+class UserWithCreditQuerySet(models.QuerySet):
```

Comment:
Django combines it into a single SQL query, so since it's a single query the database can do the same optimizations as for a SQL view which would likely use the same or similar query. I don't really mind what to do.

Some interesting Google results:
* https://dba.stackexchange.com/questions/151169/are-views-harmful-for-performance-in-postgresql
* https://stackoverflow.com/questions/1108664/difference-between-views-and-select-queries

We could possibly use views to improve performance but it will need to require extra effort. We should of course also only optimize after having profiled first.
* https://www.postgresql.org/docs/11/rules-materializedviews.html
* https://docs.djangoproject.com/en/2.1/topics/db/optimization/
* This is specific to Microsoft SQL server but still interesting: https://stackoverflow.com/questions/439056/is-a-view-faster-than-a-simple-query

Reply:
If Django combines the whole thing into one query it should be quite performant indeed. I agree, at least good enough for now, optimize later (but probably that is not necessary) :wink: